### PR TITLE
Reworks the prometheus metrics to adhere to best practices

### LINKED ISSUE #4854: Standardize prometheus naming convention for keda metrics
### Proposal

Because of issues with Prometheus remote writers, it's ideal that we follow Prometheus suffix conventions for counters. If not followed strictly, external vendors supporting Prometheus remote writers might receive the metric as a `Gauge` instead.

Context:
* https://prometheus.io/docs/practices/naming/
* https://docs.newrelic.com/docs/infrastructure/prometheus-integrations/install-configure-remote-write/set-your-prometheus-remote-write-integration/#mapping
* https://github.com/prometheus/client_golang/issues/725
* https://github.com/thanos-io/thanos/pull/3992/filesPrometheus

### Use-Case

Vendors such as newrelic can use Prometheus metrics via remote writer properly.

### Is this a feature you are interested in implementing yourself?

Maybe

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by wozniakjan**
This is continuing work from https://github.com/kedacore/keda/pull/5174. I have rebased on `main`, resolved conflicts and working on fixing the e2e bug as described in https://github.com/kedacore/keda/pull/5174#issuecomment-1956280930.

Original text from @wonko 
> I reworked the prometheus metric names to follow conventions: added units where needed, and appended `total`.
>
> Also passed a `time.Duration` down to the metrics handlers, as the conversion to units needs to happen when construction the metrics.
>
> No tests were present, none added.
>
> No idea where to track deprecation actions (when and how will the deprecated metrics be removed?)

There are two parts to this PR
* original work from [#5174](https://github.com/kedacore/keda/pull/5174) squashed and rebased on latest `main` - 6e5f45c0
* changes from this PR to ensure no original metric gets broken - 31dd0609, d2ee08f1, c60020b7, 25783ee7

### Checklist

- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) - [keda-docs#1374](https://github.com/kedacore/keda-docs/pull/1374)
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #4854
Closes https://github.com/kedacore/keda/pull/5174
Relates to https://github.com/kedacore/keda-docs/pull/1374

### CODE REVIEW

**Comment by wozniakjan** [comment]
* https://github.com/kedacore/keda/pull/5687/commits/ced62ff46994542696f002c0f046d5702830ede8 contains original work from https://github.com/kedacore/keda/pull/5174
* https://github.com/kedacore/keda/pull/5687/commits/e8d3c37f2c1b71efcc0ae65ece13283173f6c32e fixes the duplicate metric registration for `keda_scaler_error_total`
* https://github.com/kedacore/keda/pull/5687/commits/30e16b9aed6a6e54a72290776545a4184f908d5e adds back `scaled_job_errors` with proper deprecation message

**Comment by wonko** [comment]
thx for picking this up @wozniakjan

**Comment by wozniakjan** [comment]
tests finally pass, ptal @kedacore/keda-core-contributors

File: CHANGELOG.md [hunk 1]
```diff
@@ -95,7 +96,7 @@ You can find all deprecations in [this overview](https://github.com/kedacore/ked
 
 New deprecation(s):
 
-- TODO ([#XXX](https://github.com/kedacore/keda/issues/XXX))
+- Various Prometheus metrics have been renamed to follow the preferred naming conventions. The old ones are still available, but will be removed in the future ([#4854](https://github.com/kedacore/keda/issues/4854)).
```
Comment (wozniakjan) [inline CHANGELOG.md:99]: @JorTurFer, @zroubalik, @tomkerkhove I am not entirely sure this is the right place to document the deprecation, I took it from the original PR https://github.com/kedacore/keda/pull/5174.

File: CHANGELOG.md [hunk 2]
```diff
@@ -73,6 +73,7 @@ Here is an overview of all new **experimental** features:
 - **General**: Add OPENTELEMETRY flag in e2e test YAML ([#5375](https://github.com/kedacore/keda/issues/5375))
 - **General**: Add support for cross tenant/cloud authentication when using Azure Workload Identity for TriggerAuthentication ([#5441](https://github.com/kedacore/keda/issues/5441))
 - **General**: Add `validations.keda.sh/hpa-ownership` annotation to HPA to disable ownership validation ([#5516](https://github.com/kedacore/keda/issues/5516))
+- **General**: Renamed Prometheus metrics to include units and `total` where approriate ([#4854](https://github.com/kedacore/keda/issues/4854))
```
Comment (tomkerkhove) [inline CHANGELOG.md:76]:
```suggestion
- **General**: Improve Prometheus metrics to align with best practices ([#4854](https://github.com/kedacore/keda/issues/4854))
```

**Comment by tomkerkhove** [comment]
@wozniakjan It looks like https://github.com/kedacore/keda-docs/pull/1258 is not owned by you, but linked. Is this one still up to date or do we need another one?

File: CHANGELOG.md (hunk 1 above)
Comment (tomkerkhove) [reply CHANGELOG.md:99]: We'll need a new issue for this; if I understand correctly we are just renaming them and keep the old ones. Once our doc PR is ready I will create required issue/discussions

**Comment by wozniakjan** [comment]
> It looks like kedacore/keda-docs#1258 is not owned by you, but linked. Is this one still up to date or do we need another one?

correct, kedacore/keda-docs#1258 was implemented by @wonko just like the majority of this PR originates in @wonko's PR https://github.com/kedacore/keda/pull/5174

There are a couple of minor suggestions in kedacore/keda-docs#1258, perhaps I can open a new docs PR that reflects reviewer suggestions.

> We'll need a new issue for this; if I understand correctly we are just renaming them and keep the old ones.

correct, the old ones remained, just in the description KEDA will inform users these are deprecated. The new ones should be in alignment with prometheus conventions.

**Comment by zroubalik** [comment]
> There are a couple of minor suggestions in [kedacore/keda-docs#1258](https://github.com/kedacore/keda-docs/pull/1258), perhaps I can open a new docs PR that reflects reviewer suggestions.

if these suggestions are the only additions we can merge them in the existing PR

**Comment by zroubalik** [comment]
> > There are a couple of minor suggestions in [kedacore/keda-docs#1258](https://github.com/kedacore/keda-docs/pull/1258), perhaps I can open a new docs PR that reflects reviewer suggestions.
> 
> if these suggestions are the only additions we can merge them in the existing PR

oh, I can see the PR is opened against 2.13, we should move that to 2.14. @wozniakjan could you please open a new doc PR then?

**Review by zroubalik — APPROVED** [review]

**Pull request merged**